File: Src/Python/votechain.py
import hashlib
import json
from math import log2
from time import time, strptime, mktime
from uuid import uuid4
from collections import OrderedDict

class VoteChain(object):
	def __init__(self):
		self.merkle_tree = OrderedDict()
		self.chain = []
		self.current_transactions = []
		self.voters = set()
	'''
		체인의 첫번째 블록
		:param
		constructor : 첫번째 블록을 생성한 계정
		subject : 해당 블록 체인의 주제
	'''

	def genesis_block(self, constructor, subject, dl):
		dl = deadline(dl)
		if dl:
			return False

		block_header = {
			'index' : 1,
			'constructor' : constructor,
			'timestamp' : time(),
			'deadline' : dl,
		}
		block_hash = self.hash(block_header)
		block = {
			'block_hash' : block_hash,
			'block_header' : block_header,
			'subject' : subject,
		}
		self.chain.append(block)
	
	'''
		체인의 주제 반환
	'''
	def get_subject(self):
		return self.chain[0]['subject']

	'''
		블럭의 구조 정의
	'''	
	def new_block(self):
		self.merkle_tree = self.transaction_record()
		block_header = {
			'index' : len(self.chain) + 1,
			'timestamp' : time(),
			'proof' : 0, 
			'previous_hash' : self.chain[-1]['block_hash'],
			'merkle_root': self.merkle_tree[1],
		}
		block_header = self.proof_of_work(block_header)
		block_hash = self.hash(block_header)
		block = {
			'block_hash' : block_hash,
			'block_header' : block_header, 
			'transactions' : self.current_transactions,
			'merkle_tree' : self.merkle_tree,
			}
		self.current_transactions = []		
		return block
	
	'''
		chain에 새로운 block 추가
		:param
		block : 추가할 block
	'''
	def add_block(self, block):
		self.chain.append(block)

	'''
		투표 완료계정 확인
		:params
		voters : 이미 투표에 참여한 계정 집합
	'''
	def update_voters(self, voters):
		self.voters.update(voters)

	'''
		계정이 투표 완료계정 집합에 포함되어 있는지 확인
		:params
		voter : 새로운 transaction을 생성하고자 하는 계정
	'''
	def check_voters(self, voter):
		if voter in self.voters:
			return False
		return True

	'''
		새로운 transaction 생성
		:param
		voter : 투표자 계정
		candidate : 선택한 후보
	'''
	def new_transaction(self, voter, candidate):
		transac = {
			'voter' : voter,
			'choosen candidate' : candidate,
			}
		return transac

	'''
		새로운 transaction 추가
		:param
		trnasac : 새로 추가할 transaction
	'''
	def add_transaction(self, transac):
		if not check_voters(trasac['voter']):
			self.current_transactions.append(transac)

	'''
		거래내용을 sha256 해시함수를 사용해 변환
		-OrderedDict 형태로 저장되며, root node 의 index를 1로 시작하여 BFS 알고리즘에 따라 노드에 부여한 index를 key로 가짐
		:param
		merkle_tree : transaction들의 hash tree
	'''
	def transaction_record(self):
		merkle_tree = OrderedDict()
		transactions = list(self.current_transactions)
		length = len(transactions)
		dep = int(log2(length))
		nodes = pow(2,dep)
		extra_nodes = length - nodes
		non = pow(2,dep+1)
		
		for index in range(0, extra_nodes):
			left = transactions.pop(index)
			right = transactions.pop(index)
			left_hash = self.hash(left)
			right_hash = self.hash(right)
			merkle_tree[non] = left_hash
			non += 1
			merkle_tree[non] = right_hash
			non += 1
			transactions.insert(index, left_hash+right_hash)

		while True:  
			length = len(transactions)
			dep = int(log2(length))
			nodes = non = pow(2,dep)      
			
			for index in range(0, int(nodes/2)):
				left = transactions.pop(index)
				left_hash = self.hash(left)
				merkle_tree[non] = left_hash
				non += 1
				right = transactions.pop(index)
				right_hash = self.hash(right)
				merkle_tree[non] = right_hash
				non += 1
				transactions.insert(index, left_hash+right_hash)
		
			if length is 1: 
				merkle_tree[1] = self.hash(transactions)
				return merkle_tree
	'''
		블럭체인의 nonce 값을 구하는 과정. proof 값을 1씩 증가시키며 적합한 해시값을 찾을 때 까지 반복
		:param
		block_h : 블럭의 헤더정보
	'''
	def proof_of_work(self, block_h):
		while self.valid_proof(block_h) is False:
			block_h['proof'] += 1

		return block_h

	'''
		chain에 새로 추가할 block의 유효성 검증
		:param
		block : chain에 추가할 block
		chain : 새로운 block이 추가될 chain
	'''
	def valid_block(self, block, chain):
		genesis_h = chain[0]['block_header']
		last_h = self.last_block['block_header']
		block_h = block['block_header']

		if block_h['timestamp'] > genesis_h['deadline']:
			return False

		if block_h['index'] != last_h['index'] + 1:
			return False

		if not self.valid_proof(block_h):
			return False

		if block_h['previous_hash'] != hash(last_h):
			return False

		return True

	'''
		블럭체인이 유효한지 검증하는 과정
		:param
		chain : 현재 블럭들이 포함된 리스트
	'''
	def valid_chain(self, chain):
		last_block = chain[0]
		current_index = 1

		while current_index < len(chain):
			block = chain[current_index]
			block_h = block['block_header']
			print("{}".format(last_block))
			print("{}".format(block))
			print("\n------------\n")

			if block_h['previous_hash'] != hash(last_block['block_header']):
				return False

			if not self.valid_proof(block_h):
				return False

			if block_h['timestamp'] > chain[0]['block_header']['deadline']
				return False
			
			last_block = block
			current_index += 1
		
		return True

	'''
		블럭이 분기되었을 때 충돌을 해결. 연결된 블럭의 길이가 긴 쪽의 체인이 유지된다.
		:param
		chainlist : 여러 체인들이 들어있는 리스트
	'''
	def resolve_conflicts(self, chainlist):
		for new_chain in chainlist:
			max_length = len(self.chain)
			length = len(new_chain)
			if length > max_length and self.valid_chain(new_chain):
				self.chain = new_chain

	'''
		블럭의 해시값을 구하는 과정. 해시값의 첫 4자리가 '0000'이면 적합한 것으로 함
		:param
		blokc_h : 블럭의 헤더정보
	'''
	def valid_proof(self, block_h):
		guess_hash = self.hash(block_h)
		
		if guess_hash[ :4] == "0000":
			return guess_hash
		
		else:
			return False

	'''
		투표 종료 시간을 받아 epoch time으로 변환뒤 반환하는 함수
		:param
		dl : "년도 월 일 시간" 형태로 투표 종료 시간을 받음
	'''
	def deadline(self, dl):
		try:
			dl = strptime("%Y %m %d %H", dl)
			t = gmtime()
			if dl.tm_year < t.tm_year:
				return False
			if dl.tm_mon > 12 or dl.tm_mon < 1 or ((dl.tm_year == t.tm_year) and (dl.tm_mon < t.tm_mon)):
				return False
			if dl.tm_mday > 31 or dl.tm_mday < 1 or ((dl.tm_mon == t.tm_mon) and (dl.tm_mday < t.tm_mday)):
				return False
			if dl.tm_hour > 23 or dl.tm_hour < 0 or ((dl.tm_mday == t.tm_mday) and (dl.tm_hour < t.tm_hour)):
				return False
			return mktime(dl)
		except:
			return False

	'''
		인자로 넘어온 값의 sha256 해시를 반환
	'''
	@staticmethod
	def hash(key):
		key_string = json.dumps(key, sort_keys=True).encode()
		return hashlib.sha256(key_string).hexdigest()
	
	'''
		현재 체인의 끝 블록을 반환
	'''
	@property
	def last_block(self):
		return self.chain[-1]

	def test(self):
		print(self.chain)


t = VoteChain()
t.deadline = '2018 5 20 22'
print(t.deadline)
'''
chain = VoteChain()
chain.test()
chain.genesis_block('201311105', '총학선거')
chain.new_transaction('1', 'A')
chain.new_transaction('2', 'A')
chain.new_block()
chain.test()
chain.new_transaction('3', 'B')
chain.new_transaction('4', 'C')
chain.new_block()
chain.test()
'''

File: Src/Java/VoteChain.java
package votechain;

import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.LinkedHashMap;


class genesisblock_header {
	private String ver;
	private int index;
	private float time;
	private float deadline;
	public genesisblock_header(String ver, float deadline) {
		super();
		this.ver = ver;
		this.index = 1;
		this.time = Calendar.getInstance().getTimeInMillis() / 1000;
		this.deadline = deadline;
	}
	public int getIndex() {
		return index;
	}
	public float getTime() {
		return time;
	}
	public String getVer() {
		return ver;
	}
	public float getDeadline() {
		return deadline;
	}
	@Override
	public String toString() {
		return "[ver=" + ver + ", index=" + index + ", time="
				+ time + ", deadline=" + deadline + "]";
	}
	
}

class genesisblock {
	private String block_hash;
	private String subject;
	private String constructor;
	private genesisblock_header genesis_h;

	public genesisblock(String block_hash, String constructor,String subject, genesisblock_header genesis_h) {
		super();
		this.block_hash = block_hash;
		this.subject = subject;
		this.constructor = constructor;
		this.genesis_h = genesis_h;
	}

	public String getBlock_hash() {
		return block_hash;
	}

	public String getSubject() {
		return subject;
	}
	
	public String getConstructor() {
		return constructor;
	}
	
	public genesisblock_header getGenesis_h() {
		return genesis_h;
	}

	public void setGenesis_h(genesisblock_header genesis_h) {
		this.genesis_h = genesis_h;
	}

	@Override
	public String toString() {
		return "[genesis_header=" + genesis_h.toString() + ", block_hash=" + block_hash + ", subject=" + subject + ", constructor=" + constructor
				+ "]";
	}
	
	
}

class block_header {
	private String ver;
	private int index;
	private int proof;
	private float time;
	private String previous_hash;
	private String merkle_root;

	public block_header(String ver, int index, String previous_hash, String merkle_root) {
		super();
		this.ver = ver;
		this.index = index;
		this.proof = 0;
		this.time = Calendar.getInstance().getTimeInMillis() / 1000;
		this.previous_hash = previous_hash;
		this.merkle_root = merkle_root;
	}
	public int getIndex() {
		return index;
	}
	public String getVer() {
		return ver;
	}
	public int getProof() {
		return proof;
	}
	public float getTime() {
		return time;
	}
	public String getPrevious_hash() {
		return previous_hash;
	}
	public String getMerkle_root() {
		return merkle_root;
	}
	
	public void setMerkle_root(String merkle_root) {
		this.merkle_root = merkle_root;
	}
	public void setProof(int proof) {
		this.proof += proof;
	}
	@Override
	public String toString() {
		return "[ver=" + ver + ", index=" + index + ", proof=" + proof + ", time=" + time
				+ ", previous_hash=" + previous_hash + ", merkle_root=" + merkle_root + "]";
	}
}

class block{
	private String block_hash;
	private ArrayList<HashMap> transaction_pool;
	private HashMap<Integer, String> merkle_tree;
	private block_header block_h;

	
	public block(ArrayList<HashMap> transaction_pool, HashMap<Integer, String> merkle_tree,
			block_header block_h) {
		super();
		this.block_hash = vote_block.hash(block_h.toString());
		this.transaction_pool = transaction_pool;
		this.merkle_tree = merkle_tree;
		this.block_h = block_h;
	}
	
	public String getBlock_hash() {
		return block_hash;
	}
	public ArrayList<HashMap> getTransaction_pool() {
		return transaction_pool;
	}
	public HashMap<Integer, String> getMerkle_tree() {
		return merkle_tree;
	}
	public block_header getBlock_h() {
		return block_h;
	}

	public void setBlock_h(block_header block_h) {
		this.block_h = block_h;
	}

	@Override
	public String toString() {
		return "[block_header=" + block_h.toString() + ", block_hash=" + block_hash + ", transaction_pool=" + transaction_pool.toString() + ", merkle_tree="
				+ merkle_tree.toString() + "]";
	}
}
class vote_chain {
	ArrayList<Object> chain;
	
	public void add_genesis(genesisblock genesis) {
		this.chain.add(0, genesis);
	}
	
	public void add_block(block block) {
		this.chain.add(block);
	}
	
	public boolean valid_chain(ArrayList<Object> chain) {
		genesisblock genesis = (genesisblock) chain.get(0);
		block last_block = (block)chain.get(1);
		int current_index = 2;
		
		while(current_index < chain.size()) {
			block block = (block)chain.get(current_index);
			block_header block_h = block.getBlock_h();
			
			if(block_h.getPrevious_hash() != vote_block.hash(last_block.toString())) return false;
			if(vote_block.valid_proof(block_h) == "false") return false;
			if(block_h.getTime() > ((genesisblock_header)chain.get(0)).getDeadline()) return false;
			
			last_block = block;
			current_index += 1;
		}
		
		return true;
	}
	
	public void reslove_conflicts(ArrayList<ArrayList> chainlist) {
		for(ArrayList new_chain : chainlist) {
			int max_length = this.chain.size();
			int length = new_chain.size();
			if(length > max_length && valid_chain(new_chain)) {
				this.chain = new_chain;
			}
		}
	}
}

class vote_block {
	LinkedHashMap<Integer, String> merkle_tree;
	ArrayList<HashMap> current_transactions;
	ArrayList<String> voters;

	public void setMerkle_tree(LinkedHashMap<Integer, String> merkle_tree) {
		this.merkle_tree = merkle_tree;
	}

	public void setCurrent_transactions(ArrayList<HashMap> current_transactions) {
		this.current_transactions = current_transactions;
	}

	public void setVoters(ArrayList<String> voters) {
		this.voters = voters;
	}

	public vote_block(LinkedHashMap<Integer, String> merkle_tree,
			ArrayList<HashMap> current_transactions, ArrayList<String> voters) {
		super();
		this.merkle_tree = merkle_tree;
		this.current_transactions = current_transactions;
		this.voters = voters;
	}

	public void update_voters(String voter) {
		this.voters.add(voter);
	}
	
	public boolean check_voters(String voter) {
		for(String vote : this.voters) {
			if(vote == voter) {
				return false;
			}
		}
		return true;
	}
	
	public HashMap<String, String> new_transaction(String voter, String candidate) {
		HashMap<String, String> transac = new HashMap<String, String>();
		transac.put("voter", voter);
		transac.put("candidate", candidate);
		return transac;
	}
	
	public void add_transaction(HashMap<String, String> transac) {
		if(!check_voters(transac.get("voter"))) {
			this.current_transactions.add(transac);
			this.merkle_tree = transaction_record();
		}
	}
	
	public boolean valid_block(block block, ArrayList<Object> chain) {
		genesisblock_header genesis_h = (genesisblock_header)chain.get(0);
		block_header block_h = block.getBlock_h();
		block_header last_h = last_block(chain).getBlock_h();
		
		if(block_h.getTime() > genesis_h.getDeadline()) return false;
		if(block_h.getIndex() != last_h.getIndex()+1) return false;
		if(valid_proof(block_h) == "false") return false;
		if(block_h.getPrevious_hash() != hash(last_h.toString())) return false;
		
		return true;
	} 
	
	public block_header proof_of_work(block_header block_h) {
		if(valid_proof(block_h) == "false") {
			block_h.setProof(1);
			return null;
		}
		return block_h;
	}
	public static String valid_proof(block_header block_h) {
		String guess_hash = hash(block_h.toString());
		if(guess_hash.substring(0, 4) == "0000") 
			return guess_hash;
		else return "false";
		
	}
	
	public float deadline(int year, int month, int day, int hour) {
		Calendar c = Calendar.getInstance();
		if(year < c.get(Calendar.YEAR)) return 0;
		if(month > 12 || month < 1) return 0;
		if(day > 31 || day < 1) return 0;
		if(hour > 23 || hour < 0) return 0;
		c.set(year, month, day, hour, 0, 0);
		return c.getTimeInMillis()/1000;
	}
	
	public LinkedHashMap<Integer, String> transaction_record() {
		LinkedHashMap<Integer, String> merkle_tree = new LinkedHashMap<>();
		ArrayList<Object> transaction = (ArrayList<Object>) this.current_transactions.clone();
		int length = transaction.size();
		int dep = (int)baseLog(length, 2);
		int nodes = (int) Math.pow(2, dep);
		int extra_nodes = length - nodes;
		int non = (int) Math.pow(2, dep+1);
		for(int i = 0; i < extra_nodes; i++) {
			HashMap left = (HashMap) transaction.get(i);
			transaction.remove(i);
			HashMap right = (HashMap) transaction.get(i);
			transaction.remove(i);
			String left_hash = hash(left.toString());
			String right_hash = hash(right.toString());
			merkle_tree.put(non, left_hash);
			non += 1;
			merkle_tree.put(non, right_hash);
			non += 1;
			transaction.add(i, left_hash+right_hash);
		}
		while(true) {
			length = transaction.size();
			dep = (int)baseLog(length, 2);
			nodes = non = (int)Math.pow(2, dep);
			for(int i = 0; i < (int)nodes/2; i++) {
				HashMap left = (HashMap) transaction.get(i);
				transaction.remove(i);
				String left_hash = hash(left.toString());
				merkle_tree.put(non, left_hash);
				non += 1;
				HashMap right = (HashMap) transaction.get(i);
				transaction.remove(i);
				String right_hash = hash(right.toString());
				merkle_tree.put(i, right_hash);
				non += 1;
				transaction.add(i, left_hash+right_hash);
			}	
			if(length == 1) {
				merkle_tree.put(1, hash(transaction.get(0).toString()));
				return merkle_tree;
			}
		}
	}	
	
	public block_header calculate_proof(block_header block_h) {
		while((proof_of_work(block_h)) == null) {
			if(block_h.getMerkle_root() != this.merkle_tree.get(1)) {
				block_h.setMerkle_root(this.merkle_tree.get(1));
			}
		}
		return block_h;
	}
	
	public static double baseLog(double x, double base) {

		return Math.log10(x) / Math.log10(base);
	}

	public static String hash(String key) {
		StringBuffer hexString = new StringBuffer();
		try {
			MessageDigest digest = MessageDigest.getInstance("SHA-256");
			byte[] hash = digest.digest(key.getBytes("UTF-8"));
			
			for (int i = 0; i < hash.length; i++) {
				String hex = Integer.toHexString(0xff & hash[i]);
				if(hex.length() == 1) hexString.append('0');
				hexString.append(hex);
			}
		} catch (Exception ex) {
			throw new RuntimeException(ex);
		}
		return hexString.toString();
	}
	private block last_block(ArrayList<Object> chain) {
		return (block)chain.get(chain.size()-1);
	}

}
